pkg/config/spec.go

```diff
@@ -400,6 +438,44 @@ func ReadInstallSpecFromConfig(c *Config) (*v1.InstallSpec, error) {
 	return installSpec, nil
 }
 
+// GetSourceSize will try to gather the actual size of the source
+// Useful to create the exact size of images and by side effect the partition size
+// This helps adjust the size to be juuuuust right.
+// It can still be manually override from the cloud config by setting all values manually
+// But by default it should adjust the sizes properly
+func GetSourceSize(config *Config, source *v1.ImageSource) (int64, error) {
+	var size int64
+	var err error
+
+	switch {
+	case source.IsDocker():
+		size, err = config.ImageExtractor.GetOCIImageSize(source.Value(), config.Platform.String())
+	case source.IsDir():
+		err = fsutils.WalkDirFs(config.Fs, source.Value(), func(path string, d fs.DirEntry, err error) error {
+			if err != nil {
+				return err
+			}
+			info, err := d.Info()
+			if err != nil {
+				return err
+			}
+			size += info.Size()
+			return nil
+		})
+
+	case source.IsFile():
+		file, err := config.Fs.Stat(source.Value())
+		if err == nil {
+			size = file.Size()
+		}
+	}
+	// Normalize size to Mb before returning and add 100Mb to round the size from bytes to mb+extra files like grub stuff
+	if size != 0 {
+		size = (size / 1000 / 1000) + 100
```

Comment:
This is integer division and it could be "throwing away" more than you expect:

```
	i := 8999999
	fmt.Println(i / 1000 / 1000)
```
In the example above, the return value is just `8`, while `9` is surely a value closer to truth. Just adding 100Mb when we have thrown away `999999` won't cut it :). Btw, what's the unit of `size`? It looks like it's Terabytes since we divide twice by 1000 to get Mb. Is that right? Maybe name the variable something like `sizeTB` to make it clear.

Reply:
Sorry, it's still morning :man_facepalming: , `size` is bytes. Ignore this. Worst case scenario, the division throws away 1 Mb and then we add 100 on top. All good.

Reply:
It's Mb.

And I had so many issues with this. I tried to make them float to not lost stuff but o was getting a zero size at the end. I'm not sure what I was doing wrong here, all sizes from the size check come in bytes but our partitioner works with Mb so we need to translate it here.

I would try again today to improve this, see if I'm more smart today than yesterday 😂